### PR TITLE
Overriding processed metrics from elasticinfra metrics processor

### PR DESCRIPTION
**Pull request by gizas**
This is the PR that tries to add the additional functionality described in https://github.com/elastic/opentelemetry-lib/issues/97#issuecomment-2376625351

The PR is tested with following configuration

```yaml
        elasticinframetrics:
          add_system_metrics: true
          add_k8s_metrics: true
           drop_original: true
```

The implementation is based on the existence of otel_remapped attribute added in remapped metrics (See [otel_remapped label](https://github.com/elastic/opentelemetry-lib/blob/6d89cbad4221429570107eb4a4968cf8a2ff919f/remappers/common/const.go#L31)) If the attribute exists then the metric is kept. 

**Note:**
No generic metrics present with  drop_original: true

No generic Metrics

K8s Metrics remapped with `add_k8s_metrics: true`
add_k8s_metrics

System metrics remapped with `add_system_metrics: true`
add_system_metrics

## How to Test

0. Prereq: Create a K8s cluster with kind
1. Clone this branch
2.Edit the [manifest.yaml](https://github.com/elastic/opentelemetry-collector-components/blob/processor/elasticinframetricsprocessor/v0.10.0/distributions/elastic-components/manifest.yaml) file with the below

<details>
<summary>Manifest Yaml File to build EDOT</summary>
```yaml
dist:
  module: github.com/elastic/opentelemetry-collector-components
  name: elastic-collector-components
  description: Testing distribution to ensure Elastic's components can be used with the OCB
  version: 0.0.1
  output_path: ./_build
  otelcol_version: 0.109.0

extensions:
  - gomod: go.opentelemetry.io/collector/extension/zpagesextension v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextension v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/sumologicextension v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/otlpencodingextension v0.109.0
    import: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/otlpencodingextension
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/jaegerencodingextension v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/avrologencodingextension v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/jsonlogencodingextension v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/textencodingextension v0.109.0
    import: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/textencodingextension
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/zipkinencodingextension v0.109.0
    import: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/zipkinencodingextension

connectors:
  - gomod: github.com/elastic/opentelemetry-collector-components/connector/spanmetricsconnectorv2 v0.0.0

converters:

receivers:
  - gomod: go.opentelemetry.io/collector/receiver/nopreceiver v0.109.0
  - gomod: go.opentelemetry.io/collector/receiver/otlpreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filelogreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/httpcheckreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/jaegerreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8seventsreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sobjectsreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kubeletstatsreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/zipkinreceiver v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sclusterreceiver v0.109.0


processors:
  - gomod: github.com/elastic/opentelemetry-collector-components/processor/elasticinframetricsprocessor v0.0.0
  - gomod: github.com/elastic/opentelemetry-collector-components/processor/elastictraceprocessor v0.0.0
  - gomod: github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor v0.0.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/attributesprocessor v0.109.0
  - gomod: go.opentelemetry.io/collector/processor/batchprocessor v0.109.0
  - gomod: go.opentelemetry.io/collector/processor/memorylimiterprocessor v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/filterprocessor v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourcedetectionprocessor v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/resourceprocessor v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor v0.109.0

exporters:
  - gomod: go.opentelemetry.io/collector/exporter/nopexporter v0.109.0
  - gomod: go.opentelemetry.io/collector/exporter/debugexporter v0.109.0
  - gomod: go.opentelemetry.io/collector/exporter/otlpexporter v0.109.0
  - gomod: go.opentelemetry.io/collector/exporter/otlphttpexporter v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter v0.109.0
  - gomod: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/fileexporter v0.109.0
  

# workaround known issue https://github.com/open-telemetry/opentelemetry-collector/releases/tag/v0.109.0
providers:
  - gomod: go.opentelemetry.io/collector/confmap/provider/envprovider v1.15.0
  - gomod: go.opentelemetry.io/collector/confmap/provider/fileprovider v1.15.0
  - gomod: go.opentelemetry.io/collector/confmap/provider/httpprovider v0.109.0
  - gomod: go.opentelemetry.io/collector/confmap/provider/httpsprovider v0.109.0
  - gomod: go.opentelemetry.io/collector/confmap/provider/yamlprovider v0.109.0

replaces:
  - github.com/elastic/opentelemetry-collector-components/processor/elasticinframetricsprocessor => ../processor/elasticinframetricsprocessor
  - github.com/elastic/opentelemetry-collector-components/processor/elastictraceprocessor => ../processor/elastictraceprocessor
  - github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor => ../processor/lsmintervalprocessor
  - github.com/elastic/opentelemetry-collector-components/connector/spanmetricsconnectorv2 => ../connector/spanmetricsconnectorv2

```
</details>

3. Run command `GOOS=linux GOARCH=arm64 make genelasticcol`
> This will build a new binary of EDOT collector inside _build folder
4. Change dir: `cd _build` 
5.  cp elastic-collector-components otelcontribcol
6. Add this dockerfile in your folder:

<details>
<summary>Dockerfile</summary>

```yaml
FROM alpine:latest as prep
RUN apk --update add ca-certificates

RUN mkdir -p /tmp

FROM scratch

ARG USER_UID=10001
USER ${USER_UID}

COPY --from=prep /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
COPY otelcontribcol /
EXPOSE 4317 55680 55679
ENTRYPOINT ["/otelcontribcol"]
CMD ["--config", "/etc/otel/config.yaml"]


```

</details>

7. Load your new docker image inside your kind k8s : `kind load docker-image otelcontribcol:latest`

8. Install the EDOT collector as described in https://github.com/rogercoll/opentelemetry/tree/add_onboarding_operator_values/docs/onboarding/8_16/operator

9. Change the image to:
```yaml
defaultCRConfig:
  image:
    repository: "otelcontribcol"
    tag: "latest"
```

## Screenshots

For k8s we have every configured interval (eg 20 sec ) the transformed metrics and all in kubernetes.pod datastream

Screenshot 2024-09-27 at 12 54 07 PM

### CODE REVIEW

**Comment by gizas** [comment]
@ishleenk17 do you have any ideas on how we can distinuqish two documents that come from the remapper? I mean what is the difference between a doc. that has ecs initial metrics and one that has the transformed for hostmetrics? I see that the names are almost the same

File: processor/elasticinframetricsprocessor/processor.go
Comment (rogercoll) [inline processor/elasticinframetricsprocessor/processor.go:84]: I think the "override" feature should not rely on specific metrics string prefixes like "kubernetes.pod", but be agnostic to the underlying remapped metrics. By "override", we should expect the processor to only output the remapped metrics, the input ones should be discarded. In that sense, the processor will only output "ECS" metrics.

File: processor/elasticinframetricsprocessor/processor.go
Comment (rogercoll) [inline processor/elasticinframetricsprocessor/processor.go:63]: What do you think if set as Remap output the new constructed `pmetrics`? In each iteration we can append a new scoped metrics to `newmetic` with the only remapped metrics but with the same source attributes. After the metrics iteration, we can decide whether to merge both `pmetric` or just return the newly created one `override`.

I haven't test it, but something like:
```suggestion
	newmetic := pmetric.NewMetrics()

	for i := 0; i < md.ResourceMetrics().Len(); i++ {
		resourceMetric := md.ResourceMetrics().At(i)
		rm := resourceMetric.Resource()

		rmnew := newmetic.ResourceMetrics().AppendEmpty()
		rm.Attributes().CopyTo(rmnew.Resource().Attributes())

		for j := 0; j < resourceMetric.ScopeMetrics().Len(); j++ {
			scopeMetric := resourceMetric.ScopeMetrics().At(j)
			rmscope := rmnew.ScopeMetrics().AppendEmpty()
			scopeMetric.Scope().Attributes().CopyTo(rmscope.Scope().Attributes())
			for _, r := range p.remappers {
				r.Remap(scopeMetric, rmscope.Metrics(), rm)
				if !p.cfg.Override {
					scopeMetric.CopyTo(rmscope)
				}
			}
		}
	}
	return newmetic, nil
}

```

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]: Agreed ! So I changed the implementation based on [otel_remapped label](https://github.com/elastic/opentelemetry-lib/blob/6d89cbad4221429570107eb4a4968cf8a2ff919f/remappers/common/const.go#L31)

File: processor/elasticinframetricsprocessor/config.go
Comment (ChrsMark) [inline processor/elasticinframetricsprocessor/config.go:23]: Is this setting going to be documented anywhere?

File: processor/elasticinframetricsprocessor/processor.go
Comment (ChrsMark) [inline processor/elasticinframetricsprocessor/processor.go:72]: What this comment describes is not an override action. Maybe that could be named to sth else like `drop_original` or something that clearly reflects the action here?

File: processor/elasticinframetricsprocessor/processor.go
Comment (devamanv) [inline processor/elasticinframetricsprocessor/processor.go:75]: Typo?
```suggestion
		newMetric := pmetric.NewMetrics()
```

File: processor/elasticinframetricsprocessor/processor.go
Comment (devamanv) [inline processor/elasticinframetricsprocessor/processor.go:31]: It's the same constant declared in `remappers/common/const.go`, right? I'm wondering if we could we use that one here instead of declaring it again?

File: processor/elasticinframetricsprocessor/processor.go
Comment (ishleenk17) [inline processor/elasticinframetricsprocessor/processor.go:89]: @gizas : BAsically, we go through the whole remapped logic and add the translated and original metrics to a scope metric.
After that you are checking for otel_remapped and creating a new resource/scope metric having only translated metric. 
Can't we have this check in the remap function and in case override is true add only the translated metric to the scopemetric?

Currently we create 2 resourcemetrics/scopemetrics.

File: processor/elasticinframetricsprocessor/processor.go
Comment (devamanv) [inline processor/elasticinframetricsprocessor/processor.go:78]: Although, it's a matter of choice, using `range` makes it look a little cleaner, while still be able to use the index. 
```suggestion
		for i := range md.ResourceMetrics() {
```
Also, another small suggestion; instead of i, j, l, and m, we could use the names that reflect what these indices represent to enhance readability e.g. 
```suggestion
		for resIdx := range md.ResourceMetrics() {
```

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go:92]: > Can't we have this check in the remap function and in case override is true add only the translated metric to the scopemetric?

I was trying to do this inside the remap initially. But then the drop has to be implemented in the infraprocessor side. And I believe now is clearer. 

> Can't we have this check in the remap function and in case override is true add only the translated metric to the scopemetric?

Yes this is what I did. I think it works now. I tested the hostemetrics as well and dont see any problems. Just wanted youe eyes on this to see if I miss something. Just another pair of eyes for testting

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go:31]: Could find a way tbh. Nice idea.

On the other hand, if someone wants to test this library is easier to find the value and not check the remote package

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]: thank done

File: processor/elasticinframetricsprocessor/config.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/config.go:23]: Updated Readme https://github.com/elastic/opentelemetry-collector-components/pull/147/files#diff-57ed28cac77217dd4bd0ef0633c9dd5d9294a2760cb3f7ecb22a59c199d750da

File: processor/elasticinframetricsprocessor/processor.go
Comment (devamanv) [reply processor/elasticinframetricsprocessor/processor.go:31]: I noticed that we are already importing the `remappers` package, that's why I asked if we could use that instead.

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go:72]: I dont have any preferences. @rogercoll any ideas? TBH I kept the word override that is in the title of this issue. 
(I translated like override flag, overrides the default functionality)

Comment (rogercoll) [reply processor/elasticinframetricsprocessor/processor.go:72]: +1 on the `drop_original` naming, it looks more descriptive.  I just changed the issue's title with the "drop" keyword instead of "override" https://github.com/elastic/opentelemetry-lib/issues/97#issuecomment-2376625351  
Thanks!

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]: Done ! thanks you !!! let me know if is ok now

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]: drop_original it is then ! I add it and docs updated

Comment (ChrsMark) [reply processor/elasticinframetricsprocessor/processor.go]: Thank's!

File: processor/elasticinframetricsprocessor/processor.go
Comment (ChrsMark) [inline processor/elasticinframetricsprocessor/processor.go:74]: Will the logic be covered by some unit tests? Even if this component will be eventually be removed when the time will be to only support otel native data, however for the following releases it seems that we will be relying on this hence it would make a lot of sense to cover the implementation with some unit tests to ensure the functionality isn't/won't be broken.

File: processor/elasticinframetricsprocessor/processor.go
Comment (devamanv) [reply processor/elasticinframetricsprocessor/processor.go]: Thanks for making the change. I think we can simply range over the returned value of `md.ResourceMetrics()`. Currently, it ranges over len. WDYT?

File: distributions/elastic-components/sample_manifest.yaml
Comment (rogercoll) [inline distributions/elastic-components/sample_manifest.yaml:1]: Is this file needed? Maybe leftover?

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go:74]: Added in https://github.com/elastic/opentelemetry-collector-components/pull/147/commits/1af050b858797a5424899d21712cdaebcee3c85c

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]:
```bash
cannot range over md.ResourceMetrics() (value of type pmetric.ResourceMetricsSlice)
```

So not possible :)

File: processor/elasticinframetricsprocessor/processor.go
Comment (rogercoll) [inline processor/elasticinframetricsprocessor/processor.go:90]: Wdyt if we use `pmetric` constants for this comparison?
```suggestion
					if metric.Type() == pmetric.MetricTypeGauge {
```

File: processor/elasticinframetricsprocessor/processor.go
Comment (rogercoll) [inline processor/elasticinframetricsprocessor/processor.go:98]:
```suggestion
					} else if metric.Type() == pmetric.MetricTypeSum {
```

File: processor/elasticinframetricsprocessor/processor.go
Comment (rogercoll) [inline processor/elasticinframetricsprocessor/processor.go:31]:
```suggestion
const OTelRemappedLabel = common.OTelRemappedLabel
```

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (rogercoll) [inline processor/elasticinframetricsprocessor/processor_test.go:136]: Shouldn't this attribute be added by opentelemetry-lib?

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (rogercoll) [inline processor/elasticinframetricsprocessor/processor_test.go:147]: If we drop the initial metrics, why are we expecting the same initial naming? Shouldn't it be `kubernetes.pod.cpu.usage.limit.pct`?

**Review by rogercoll — COMMENTED** [review]
Overall lgtm, just added some nits/questions

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor_test.go:147]: TBH I had the same question ! I will have to investigate for this and see why it does not rename/ remap it. Le you know

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (ChrsMark) [inline processor/elasticinframetricsprocessor/processor_test.go:134]: Shouldn't we have 3 types of metrics here?

- metrics that are not mapped at all and hence are dropped.
- original metrics that are mapped but are dropped.
- remapped metrics that are kept.

File: distributions/elastic-components/sample_manifest.yaml
Comment (gizas) [reply distributions/elastic-components/sample_manifest.yaml]: Removed. Was added my me

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]: Sorry this does not work!

Comment (rogercoll) [reply processor/elasticinframetricsprocessor/processor.go]: What do you mean? Are not able to import `github.com/elastic/opentelemetry-lib/remappers/common`?

Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]: It was not there this lib. I imported and now all ok!

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor_test.go:126]: I added this

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor_test.go]: I added `k8s.volume.capacity`

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor_test.go:135]: I think that two first categories are the same eventually and we dont care for the `metrics that are not mapped at all and hence are dropped.`

At the moment I dont see the remapper working correctly calling the external lib. It was not working also [here](https://github.com/elastic/opentelemetry-collector-components/blob/49e16899f84cf4e490c961d2fd38c922fc773558/processor/elasticinframetricsprocessor/processor_test.go#L55)  I think.

I will keep troubleshooting

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor_test.go:192]: @rogercoll I think with yours PR or my fix here we are ok

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor_test.go]: Indeed was not needed. I removed it

File: processor/elasticinframetricsprocessor/processor_test.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor_test.go:137]: @ChrsMark tests finally are fixed. Let me know

**Comment by rogercoll** [comment]
Follow-up task https://github.com/elastic/opentelemetry-lib/issues/103

File: processor/elasticinframetricsprocessor/testdata/k8smetrics_drop_original/output-metrics.yaml
Comment (ChrsMark) [inline processor/elasticinframetricsprocessor/testdata/k8smetrics_drop_original/output-metrics.yaml:29]: Are those `0`s here on purpose? Seems a bit odd when the input file provides the `k8s.pod.cpu_limit_utilization` equal to `345` to have additional metrics reported with zero values.

Comment (ChrsMark) [reply processor/elasticinframetricsprocessor/testdata/k8smetrics_drop_original/output-metrics.yaml:29]: Ah that's exactly what https://github.com/elastic/opentelemetry-collector-components/pull/147#issuecomment-2386081423 reports 👍🏽

**Review by ChrsMark — APPROVED** [review]
LGTM

File: processor/elasticinframetricsprocessor/processor.go
Comment (gizas) [reply processor/elasticinframetricsprocessor/processor.go]: Done in https://github.com/elastic/opentelemetry-collector-components/pull/147/files#diff-d8a33fa92f13418d93d8e208450066b53fe923e4026cc3d184eb3f89bc8d7c62R32

**Review by rogercoll — APPROVED** [review]

**Comment by gizas** [comment]
@ishleenk17 or @devamanv l need one more review from `elastic/obs-infraobs-integrations` team to merge.
Let me know if you have any more suggestions here

**Review by devamanv — APPROVED** [review]

**Pull request merged**
